File: app/src/main/java/org/wen/gank/CategoryFragment.java
package org.wen.gank;

import android.content.Context;
import android.os.Bundle;
import android.support.v4.widget.SwipeRefreshLayout;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.view.View;
import android.widget.Toast;

import org.wen.gank.api.GankApi;
import org.wen.gank.api.HttpResult;
import org.wen.gank.model.GankModel2;
import org.wen.gank.mvp.MvpFragment;
import org.wen.gank.tools.AppDatabase;
import org.wen.gank.tools.LoadMoreDelegate;
import org.wen.gank.widgets.DividerItemDecoration;

import java.util.ArrayList;
import java.util.List;

import javax.inject.Inject;

import butterknife.BindColor;
import butterknife.BindDimen;
import butterknife.BindView;
import butterknife.ButterKnife;
import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.functions.Consumer;
import io.reactivex.functions.Function;
import io.reactivex.schedulers.Schedulers;
import me.drakeet.multitype.MultiTypeAdapter;
import timber.log.Timber;

/**
 * created by Jiahui.wen 2017-07-23
 */
public class CategoryFragment extends MvpFragment<CategoryView, CategoryPresenter> implements CategoryView {

    private static final String EXTRA_CATEGORY = "category";

    public static final String ANDROID = "Android";
    public static final String IOS = "iOS";
    public static final String FRONT = "前端";

    public static CategoryFragment getInstance(String category) {
        CategoryFragment fragment = new CategoryFragment();
        Bundle data = new Bundle();
        data.putString(EXTRA_CATEGORY, category);
        fragment.setArguments(data);
        return fragment;
    }

    @Inject
    GankApi gankApi;
    @Inject
    AppDatabase mDatabase;

    @BindView(R.id.recyclerView)
    RecyclerView recyclerView;
    @BindView(R.id.swipeRefreshLayout)
    SwipeRefreshLayout swipeRefreshLayout;

    @BindColor(R.color.divider)
    int dividerColor;
    @BindDimen(R.dimen.divider)
    int dividerHeight;

    private String category;
    private int pageStart = 1;

    MultiTypeAdapter adapter;
    LoadMoreDelegate loadMoreDelegate;

    @Override
    public void onAttach(Context context) {
        super.onAttach(context);
        App.from(context).getAppComponent().inject(this);
        category = getArguments().getString(EXTRA_CATEGORY, "");
    }

    @Override
    protected CategoryPresenter createPresenter() {
        return new CategoryPresenter();
    }

    @Override
    protected int getLayoutRes() {
        return R.layout.fragment_category;
    }

    @Override
    protected void initView(View view) {
        ButterKnife.bind(this, view);
        recyclerView.setLayoutManager(new LinearLayoutManager(getContext()));
        recyclerView.addItemDecoration(new DividerItemDecoration(dividerColor, dividerHeight));
        adapter = new MultiTypeAdapter(new ArrayList<>());
        adapter.register(GankModel2.class, new CategoryItemProvider());
        recyclerView.setAdapter(adapter);
        loadMoreDelegate = new LoadMoreDelegate().adapter(adapter).listen(new LoadMoreDelegate.OnLoadMoreListener() {
            @Override
            public void onLoadMore() {
                loadData(pageStart);
            }
        }).into(recyclerView);

        swipeRefreshLayout.setOnRefreshListener(new SwipeRefreshLayout.OnRefreshListener() {
            @Override
            public void onRefresh() {
                loadData(1);
            }
        });
    }

    @Override
    protected void onLazyLoad() {
        super.onLazyLoad();

        mDatabase.gankDao().getGanksLimit(category).take(1).observeOn(AndroidSchedulers.mainThread())
                .subscribe(new Consumer<List<GankModel2>>() {
                    @Override
                    public void accept(List<GankModel2> ganks) throws Exception {
                        adapter.setItems(ganks);
                        adapter.notifyDataSetChanged();
                        swipeRefreshLayout.setRefreshing(true);
                        loadData(1);
                        Timber.d("load database cache with size: %d", ganks.size());
                    }
                });
    }

    private void loadData(final int start) {
        gankApi.getCategoryDatas(category, start, 15).subscribeOn(Schedulers.io())
                .map(new Function<HttpResult<List<GankModel2>>, List<GankModel2>>() {
                    @Override
                    public List<GankModel2> apply(HttpResult<List<GankModel2>> result) throws Exception {
                        return result.results;
                    }
                }).doOnNext(new Consumer<List<GankModel2>>() {
            @Override
            public void accept(List<GankModel2> gankModel2s) throws Exception {
                mDatabase.gankDao().batchinsert(gankModel2s);
            }
        }).observeOn(AndroidSchedulers.mainThread()).subscribe(new Consumer<List<GankModel2>>() {
            @Override
            public void accept(List<GankModel2> ganks) throws Exception {
                if (start == 1) pageStart = 1;
                pageStart++;
                if (ganks.isEmpty()) {
                    loadMoreDelegate.loadEnd();
                } else {
                    loadMoreDelegate.loadFinish(ganks, start == 1);
                }
                swipeRefreshLayout.setRefreshing(false);
            }
        }, new Consumer<Throwable>() {
            @Override
            public void accept(Throwable throwable) throws Exception {
                Toast.makeText(getContext(), "Error", Toast.LENGTH_SHORT).show();
                swipeRefreshLayout.setRefreshing(false);
                loadMoreDelegate.loadError(throwable);
            }
        });
    }
}

File: app/src/main/java/org/wen/gank/CategoryFragment.kt
package org.wen.gank

import android.content.Context
import android.os.Bundle
import android.view.View
import android.widget.Toast
import androidx.recyclerview.widget.LinearLayoutManager
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.disposables.CompositeDisposable
import io.reactivex.schedulers.Schedulers
import kotlinx.android.synthetic.main.fragment_category.*
import me.drakeet.multitype.MultiTypeAdapter
import org.koin.android.ext.android.inject
import org.wen.gank.api.GankApi
import org.wen.gank.extensions.getCompatColor
import org.wen.gank.extensions.getDimen
import org.wen.gank.extensions.plusAssign
import org.wen.gank.model.GankModel
import org.wen.gank.mvp.MvpFragment
import org.wen.gank.tools.AppDatabase
import org.wen.gank.tools.LoadMoreDelegate
import org.wen.gank.widgets.DividerItemDecoration
import timber.log.Timber
import java.util.*

/**
 * created by Jiahui.wen 2017-07-23
 */
class CategoryFragment : MvpFragment<CategoryView, CategoryPresenter>(), CategoryView {

  private val gankApi: GankApi by inject()
  private val mDatabase: AppDatabase by inject()

  override var layoutRes: Int = R.layout.fragment_category

  private var dividerColor: Int = 0
  private var dividerHeight: Int = 0

  private lateinit var category: String
  private var pageStart = 1

  private lateinit var adapter: MultiTypeAdapter
  private lateinit var loadMoreDelegate: LoadMoreDelegate
  private lateinit var disposes: CompositeDisposable

  override fun onAttach(context: Context) {
    super.onAttach(context)
    category = arguments?.getString(EXTRA_CATEGORY, ANDROID) ?: ANDROID
    dividerColor = context.getCompatColor(R.color.divider)
    dividerHeight = context.getDimen(R.dimen.divider)
  }

  override fun createPresenter(): CategoryPresenter {
    return CategoryPresenter()
  }

  override fun initView(view: View) {
    recyclerView.layoutManager = LinearLayoutManager(context)
    recyclerView.addItemDecoration(DividerItemDecoration(dividerColor, dividerHeight))
    adapter = MultiTypeAdapter(ArrayList<Any>())
    adapter.register(GankModel::class.java, CategoryItemProvider())
    recyclerView.adapter = adapter
    loadMoreDelegate = LoadMoreDelegate().adapter(adapter).listen(object : LoadMoreDelegate.OnLoadMoreListener {
      override fun onLoadMore() {
        loadData(pageStart)
      }
    }).into(recyclerView!!)

    swipeRefreshLayout.setOnRefreshListener { loadData(1) }

    disposes = CompositeDisposable()
  }

  override fun onLazyLoad() {
    super.onLazyLoad()

    val dispose = mDatabase.gankDao().getGanksLimit(category).take(1).observeOn(AndroidSchedulers.mainThread())
        .subscribe { ganks ->
          adapter.items = ganks
          adapter.notifyDataSetChanged()
          swipeRefreshLayout!!.isRefreshing = true
          loadData(1)
          Timber.d("load database cache with size: %d", ganks.size)
        }
    disposes += dispose
  }

  private fun loadData(start: Int) {
    val dispose = gankApi.getCategoryDatas(category, start, 15).subscribeOn(Schedulers.io())
        .map { (_, results) -> results }.doOnNext { gankModels -> mDatabase.gankDao().batchInsert(gankModels) }.observeOn(AndroidSchedulers.mainThread()).subscribe({ ganks ->
          if (start == 1) pageStart = 1
          pageStart++
          if (ganks.isEmpty()) {
            loadMoreDelegate.loadEnd()
          } else {
            loadMoreDelegate.loadFinish(ganks, start == 1)
          }
          swipeRefreshLayout.isRefreshing = false
        }, { throwable ->
          Toast.makeText(context, "Error", Toast.LENGTH_SHORT).show()
          swipeRefreshLayout!!.isRefreshing = false
          loadMoreDelegate.loadError(throwable)
          Timber.e(throwable)
        })
    disposes += dispose
  }

  override fun onDestroyView() {
    super.onDestroyView()
    disposes.clear()
  }

  companion object {

    private const val EXTRA_CATEGORY = "category"

    const val ANDROID = "Android"
    const val IOS = "iOS"
    const val FRONT = "前端"

    fun getInstance(category: String): CategoryFragment {
      val fragment = CategoryFragment()
      val data = Bundle()
      data.putString(EXTRA_CATEGORY, category)
      fragment.arguments = data
      return fragment
    }
  }
}
